drafts/when-elixirs-performance-becomes-rust-y.md

```diff
@@ -0,0 +1,405 @@
+
+## Starting off with pure Elixir
+
+So let's get coding! We'll start off with the Elixir implementation after making a new project. In this example the project is called `rust_nif`.
+
+```Elixir
+defmodule RustNif.ElixirPrimes do
+  def prime_numbers(numbers) do
+    prime_numbers(numbers, [])
+  end
+
+  def prime_numbers([], result) do
+    {:ok, result |> Enum.reverse() }
+  end
+
+  def prime_numbers([number | rest], result) do
+    new_result = result
+    |> add_if_prime_number(number)
+
+    prime_numbers(rest, new_result)
+  end
+
+  defp add_if_prime_number(numbers, 1), do: numbers
+
+  defp add_if_prime_number(numbers, 2) do
+    [2 | numbers]
+  end
+
+  defp add_if_prime_number(numbers, n) do
+    range = 2..(n - 1)
+    case Enum.any?(range, fn x -> rem(n, x) == 0 end) do
+      false -> [n | numbers]
+      _ -> numbers
+    end
+  end
+end
+
+
+time = Time.utc_now
+RustNif.ElixirPrimes.prime_numbers(Enum.into 1..100000, []) |> IO.inspect
+IO.puts "Elixir task finished after #{Time.diff Time.utc_now, time} seconds"
+```
+
+We prepend each number to the list, and reverse the list in the final result. See [this section](http://erlang.org/doc/efficiency_guide/listHandling.html) as to why we do this.
+Running this, the results on my machine were:
+
+```
+{:ok,
+ [2, 3, 5, 7, 11, 13, 17, 19, 23, 29, 31, 37, 41, 43, 47, 53, 59, 61, 67, 71,
+  73, 79, 83, 89, 97, 101, 103, 107, 109, 113, 127, 131, 137, 139, 149, 151,
+  157, 163, 167, 173, 179, 181, 191, 193, 197, 199, 211, 223, ...]}
+Elixir task finished after 31 seconds
+```
+
+That takes quite some time. Surely we can do better.
+
+## So what are our alternatives?
+
+### Ports & Nifs
+
+We could resort to ports or NIFs, if you would want to improve this by writing it in another language.
+There's a good summary of ports and nifs here: https://spin.atomicobject.com/2015/03/16/Elixir-native-interoperability-ports-vs-nifs/
+
+In short:
+
+Ports:
++ Safety
++ Error trapping
++ Flexible communication
++ No external Erlang/Elixir specific libraries required
+- Communication via STDIN and STDOUT
+
+NIFS:
++ Fast and simple implementationn
++ No context switch required
++ Simple debugging
+- Not very safe (a bug in your NIF could crash the whole VM)
+- NIF-specific C libraries
+- Native implementation can leak into Elixir code
+
+### (Web) service
+
+We could define a separate web service, exposing an API that could handle this, in a language better suited for computation heavy tasks. We'd need to add a new service to our stack though, and also account for potential network failures and sending potentially large payloads.
+
+### Background job
+
+Depending on the case, it would be acceptable to move this computation into a background task. But in our scenario, this isn't the case
```

Comment:
What do you mean by background job? ErlangVM runs a gazillion processes at a time, there is no 'concept' of separate background jobs (like with Rails/Sidekiq) ;)